### PR TITLE
Add fast shard pruning path for INSERTs on hash partitioned tables

### PR DESCRIPTION
**Pull request by onderkalaci**
This commit adds a fast shard pruning path for INSERTs on
hash-partitioned tables. The rationale behind this change is
that if there exists a sorted shard interval array, a single
index lookup on the array allows us to find the corresponding
shard interval. As mentioned above, we need a sorted
(wrt shardminvalue) shard interval array. Thus, this commit
updates shardIntervalArray to sortedShardIntervalArray in the
metadata cache. Then uses the low-level API that is defined in
multi_copy to handle the fast shard pruning.

The performance impact of this change is more apparent as more
shards exist for a distributed table. Previous implementation
was relying on linear search through the shard intervals. However,
this commit relies on constant lookup time on shard interval
array. Thus, the shard pruning becomes less dependent on the
shard count.

### CODE REVIEW

**Comment by onderkalaci** [comment]
I remember @sumedhpathak mentioned that @anarazel will review that. @sumedhpathak If I'm wrong, could you assign the reviewer? Thanks!

**Comment by onderkalaci** [comment]
@anarazel I chatted with @sumedhpathak and he said you're the reviewer for this PR.

Thanks!

File: src/backend/distributed/planner/modify_planner.c [hunk 1]
```diff
@@ -544,6 +568,86 @@ DistributedModifyShardInterval(Query *query)
...
+static bool
+UseFastShardPruning(CmdType commandType, char partitionMethod)
+{
+	/* we currently only support INSERTs */
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	/* INSERTs could only go into hash and range distributed tables */
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: Easy to misunderstand comment - if that's the case, there's no need for an if... Maybe say that's currently the case?

File: src/backend/distributed/planner/modify_planner.c [hunk 2]
```diff
@@ -544,6 +568,86 @@ DistributedModifyShardInterval(Query *query)
...
+	bool useBinarySearch = false;
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+	char partitionMethod = PartitionMethod(distributedTableId);
+	FmgrInfo *shardIntervalCompareFunction = NULL;
+	FmgrInfo *hashFunction = NULL;
+	ShardInterval *shardInterval = NULL;
+
+	/* determine whether to use binary search */
+	if (PartitionMethod(distributedTableId) != DISTRIBUTE_BY_HASH ||
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: You've already determined the partition method above.

More generally I think it's a bad idea to do a cache lookup for DistributedTableCacheEntry(), PartitionColumn(), PartitionMethod() and PartitionMethod() again. Given that you've looked up the cache entry, just use partitionEntry->partitionMethod instead of calling PartitionMethod() which does a separate hash lookup.

File: src/backend/distributed/planner/modify_planner.c [hunk 3]
```diff
@@ -544,6 +568,86 @@ DistributedModifyShardInterval(Query *query)
...
+	if (useBinarySearch || partitionMethod != DISTRIBUTE_BY_HASH)
+	{
+		shardIntervalCompareFunction = LookupShardIntervalCompareFunction(partitionColumn,
+																		  partitionMethod);
+	}
+
+	/* we only need hash functions for hash distributed tables */
+	if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		TypeCacheEntry *typeEntry = lookup_type_cache(partitionColumn->vartype,
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: I'm fine with writing this of as a separate optimization, but for performance I think it'd be a rather good idea to cache these in the DistTableCacheEntry.

File: src/backend/distributed/planner/modify_planner.c [hunk 4]
```diff
@@ -507,16 +511,16 @@ DistributedModifyJob(Query *query, Task *modifyTask)
...
 {
-	List *restrictClauseList = NIL;
 	List *prunedShardList = NIL;
-	Index tableId = 1;
 
+	int shardCount = 0;
 	Oid distributedTableId = ExtractFirstDistributedTableId(query);
-	List *shardIntervalList = NIL;
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	char partitionMethod = PartitionMethod(distributedTableId);
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: Same thing about repeated cache lookups again.

File: src/backend/distributed/planner/modify_planner.c [hunk 5]
```diff
@@ -528,9 +532,29 @@ DistributedModifyShardInterval(Query *query)
...
 	}
 
-	restrictClauseList = QueryRestrictList(query);
-	prunedShardList = PruneShardList(distributedTableId, tableId, restrictClauseList,
-									 shardIntervalList);
+
+	if (UseFastShardPruning(query->commandType, partitionMethod))
+	{
+		uint32 rangeTableId = 1;
+		Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: And here again.

File: src/backend/distributed/planner/modify_planner.c [hunk 6]
```diff
@@ -544,6 +568,86 @@ DistributedModifyShardInterval(Query *query)
...
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+	char partitionMethod = PartitionMethod(distributedTableId);
+	FmgrInfo *shardIntervalCompareFunction = NULL;
+	FmgrInfo *hashFunction = NULL;
+	ShardInterval *shardInterval = NULL;
+
+	/* determine whether to use binary search */
+	if (PartitionMethod(distributedTableId) != DISTRIBUTE_BY_HASH ||
+		!IsUniformHashDistribution(sortedShardIntervalArray, shardCount))
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: Shouldn't we cache IsUniformHashDistribution's result? It's exactly free to iterate through all shards?

File: src/backend/distributed/utils/metadata_cache.c [hunk 7]
```diff
@@ -89,6 +92,43 @@ IsDistributedTable(Oid relationId)
 
 
 /*
+ * HasUninitializedShardInterval returns true if the distributed relation has
+ * at least one shard interval whose min/max value has not been set.
+ */
+bool
+HasUninitializedShardInterval(Oid relationId)
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: It seems better to store this as an attribute to DistTableCacheEntry. E.g. DistTableCacheEntry->shardIntervalHasUninitialized (and then DistTableCacheEntry->shardIntervalIsUniform).

File: src/backend/distributed/utils/metadata_cache.c [hunk 8]
```diff
@@ -268,6 +312,31 @@ LookupDistTableCacheEntry(Oid relationId)
 		heap_close(distShardRelation, AccessShareLock);
 	}
 
+	/* sort the interval array */
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: Maybe put that in a helper function instead? Possibly with more related code?

File: src/backend/distributed/utils/shardinterval_utils.c [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+ *
+ * If a shard interval does not have min/max value, it's treated as being greater
+ * than the other.
+ */
+int
+CompareShardIntervals(const void *leftElement, const void *rightElement,
+					  FmgrInfo *typeCompareFunction)
+{
+	ShardInterval **leftShardInterval = (ShardInterval **) leftElement;
+	ShardInterval **rightShardInterval = (ShardInterval **) rightElement;
```
Comment (anarazel) [inline src/backend/distributed/utils/shardinterval_utils.c]: Given that you're only dealing with dereferenced ShardInterval **'s in the rest of the function, maybe dereference them here? I think that'd make the code easier to read.

File: src/backend/distributed/utils/shardinterval_utils.c [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+	{
+		comparisonResult = -1;
+		return comparisonResult;
+	}
+
+	/* if both shard interval have min/max values, calculate the comparison result */
+	leftDatum = (*leftShardInterval)->minValue;
+	rightDatum = (*rightShardInterval)->minValue;
+
+	Datum comparisonDatum = CompareCall2(typeCompareFunction, leftDatum, rightDatum);
```
Comment (anarazel) [inline src/backend/distributed/utils/shardinterval_utils.c]: This will warn in some compilers, because you're declaring a variable after code.

File: src/backend/distributed/utils/shardinterval_utils.c [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+	{
+		ShardIntervalCompareFunctionCacheEntry *cacheEntry = lfirst(cacheEntryCell);
+
+		if ((equal(partitionColumn, cacheEntry->partitionColumn)) &&
+			(cacheEntry->partitionMethod == partitionMethod))
+		{
+			functionInfo = cacheEntry->functionInfo;
+			break;
+		}
+	}
```
Comment (anarazel) [inline src/backend/distributed/utils/shardinterval_utils.c]: I think it's a pretty bad idea to add further caches of this kind. Using O(n) datastructures for caching O(log(n)) accesses is a seriously bad idea.  Why don't we just instead store the interval comparison function in the table cache?

Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: Do you think it's ok to add more stuff to the metadata cache? I thought we only keep copies of the metadata tables in the cache. But, if you think it's ok to add these functions to cache, it's OK for me as well.

Comment (anarazel) [inline src/backend/distributed/utils/shardinterval_utils.c]: > > +	ShardIntervalCompareFunctionCacheEntry *matchingCacheEntry = NULL;
> > +	ListCell *cacheEntryCell = NULL;
> > +	FmgrInfo *functionInfo = NULL;
> > +
> > +	/* search the cache */
> > +	foreach(cacheEntryCell, ShardIntervalCompareFunctionCache)
> > +	{
> > +		ShardIntervalCompareFunctionCacheEntry *cacheEntry = lfirst(cacheEntryCell);
> > +
> > +		if ((equal(partitionColumn, cacheEntry->partitionColumn)) &&
> > +			(cacheEntry->partitionMethod == partitionMethod))
> > +		{
> > +			functionInfo = cacheEntry->functionInfo;
> > +			break;
> > +		}
> > +	}
> 
> Do you think it's ok to add more stuff to the metadata cache?

Why not?


> I thought we only keep copies of the metadata tables in the cache.

Information about a table's partition column are that imo. And the
metadata cache handles invalidation and everything, which your
hand-rolled cached doesn't...

Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: Agreed, I'll add those to the metadata cache, which will potentially simplify the PR by removing some codes.

File: src/include/distributed/metadata_cache.h [hunk 12]
```diff
@@ -38,11 +38,12 @@ typedef struct
 
 	/* pg_dist_shard metadata (variable-length ShardInterval array) for this table */
 	int shardIntervalArrayLength;
-	ShardInterval *shardIntervalArray;
+	ShardInterval **sortedShardIntervalArray;
```
Comment (onderkalaci) [inline src/include/distributed/metadata_cache.h]: I have a question here. 

I don't see any drawbacks of keeping an array of shard interval pointers instead of array of shard intervals. But, do you think are there any drawbacks that I might have missed?

Comment (anarazel) [inline src/include/distributed/metadata_cache.h]: I was wondering about this actually. Why did you change this? To make the sorting cheaper?   There's a disadvantage: It makes binary search slower. Direct value lookups are less expensive than lookups via an indirection (pointer).

Comment (onderkalaci) [inline src/include/distributed/metadata_cache.h]: In the code, whenever we needed shard interval arrays, we use array of shard interval pointers instead of array of shard intervals. So, having that in the cache makes its utilization in the code smoother. I think having slower binary search is an acceptable trade of. Is that OK for you?

**Comment by onderkalaci** [comment]
@anarazel I applied your feedback. I added 4 more fields to the disttablecache (hash function, shard interval compare function, hasUninitiliazedShardInterval, hasUniformHashDistribution). Also, removed repetitive calls to cache entry via extra function calls. I only kept calls to `PartitionColumn()` since it isn't in the cache as it is and requires some function calls. Is that OK?

File: src/backend/distributed/utils/shardinterval_utils.c (hunk 10 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: fixed

File: src/backend/distributed/utils/shardinterval_utils.c (hunk 9 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: done

File: src/backend/distributed/utils/shardinterval_utils.c (hunk 10 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: done

File: src/backend/distributed/utils/shardinterval_utils.c (hunk 9 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c]: Done.

File: src/backend/distributed/utils/metadata_cache.c (hunk 8 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: Yes, that makes sense. Moved to a function

File: src/backend/distributed/utils/metadata_cache.c (hunk 7 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: added to the cache.

File: src/backend/distributed/planner/modify_planner.c (hunk 6 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/modify_planner.c]: Yeah it is an expensive function especially for shard count >> 1024. So, added to cache.

File: src/backend/distributed/planner/modify_planner.c (hunk 3 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/modify_planner.c]: agreed, it looks better in the metadata cache.

File: src/backend/distributed/utils/metadata_cache.c [hunk 13]
```diff
@@ -636,11 +840,19 @@ ResetDistTableCacheEntry(DistTableCacheEntry *cacheEntry)
...
-		cacheEntry->shardIntervalArray = NULL;
+		pfree(cacheEntry->sortedShardIntervalArray);
+		cacheEntry->sortedShardIntervalArray = NULL;
 		cacheEntry->shardIntervalArrayLength = 0;
+
+		cacheEntry->hasUninitializedShardInterval = false;
+		cacheEntry->hasUniformHashDistribution = false;
+
+		pfree(cacheEntry->shardIntervalCompareFunction);
+		cacheEntry->shardIntervalCompareFunction = NULL;
```
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: One question here, do I need to free hashFunction as well? I think I shouldn't be doing it since we're using typecache enrty. Is that right?

File: src/include/distributed/metadata_cache.h (hunk 12 above)
Comment (anarazel) [inline src/include/distributed/metadata_cache.h]: > > @@ -38,11 +38,12 @@ typedef struct
> >  
> >  	/* pg_dist_shard metadata (variable-length ShardInterval array) for this table */
> >  	int shardIntervalArrayLength;
> > -	ShardInterval *shardIntervalArray;
> > +	ShardInterval **sortedShardIntervalArray;
> 
> In the code, whenever we needed shard interval arrays, we use array of
> shard interval pointers instead of array of shard intervals.

Well, I tend to agree that that's a mistake. There's, based on a quick
look, no reason for doing so. And in many of these cases we access all
of them, and there's often a good number. So spatial locality does play
a role.

> So, having that in the cache makes its utilization in the code
> smoother. I think having slower binary search is an acceptable trade
> of. Is that OK for you?

The above said, I think it's mostly a minor detail. So I think you can
decide where to go.

**Comment by onderkalaci** [comment]
@anarazel rebased and pushed back

**Comment by onderkalaci** [comment]
@anarazel rebased and push back once more :)

File: src/backend/distributed/utils/metadata_cache.c [hunk 14]
```diff
@@ -286,14 +352,152 @@ LookupDistTableCacheEntry(Oid relationId)
...
+	return shardIntervalCompareFunction;
+}
+
+
+/*
+ * SortedShardIntervalArray sorts the input shardIntervalArray. Shard intervals with
+ * no min/max values are placed at the end of the array.
+ */
+static ShardInterval **
+SortedShardIntervalArray(ShardInterval **shardIntervalArray, int shardCount,
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: Shouldn't this be SortShardInterval()?

File: src/backend/distributed/planner/modify_planner.c [hunk 15]
```diff
@@ -543,6 +569,74 @@ DistributedModifyShardInterval(Query *query)
 
 
 /*
+ * UseFastShardPruning returns true if the commandType is INSERT and partition method
+ * is hash or range.
+ */
+static bool
+UseFastShardPruning(CmdType commandType, char partitionMethod)
```
Comment (anarazel) [inline src/backend/distributed/planner/modify_planner.c]: This sounds more like a an action than a boolean function. Maybe FastShardPruningPossible()?

File: src/backend/distributed/utils/metadata_cache.c (hunk 13 above)
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: I think it'd be better to copy the typecache's fmgr_info - the two caches don't necessarily have the same lifetimes.

File: src/backend/distributed/planner/modify_planner.c (hunk 15 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/modify_planner.c]: That makes sense.

File: src/backend/distributed/utils/metadata_cache.c (hunk 14 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: Yeap, I used `SortShardIntervalArray()`

File: src/backend/distributed/utils/metadata_cache.c (hunk 13 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: Ok, I allocated the hash function from the Cache context and released as the others.

**Comment by onderkalaci** [comment]
applied your feedback and re-based latest changes.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 16]
```diff
@@ -2056,12 +1989,11 @@ SubquerySqlTaskList(Job *job)
 			return NIL;
 		}
 
-		sortedIntervalArray = SortedShardIntervalArray(finalShardIntervalList);
 		fragmentCombinationCell = list_head(fragmentCombinationList);
 
 		for (shardIndex = 0; shardIndex < finalShardCount; shardIndex++)
 		{
-			ShardInterval *shardInterval = sortedIntervalArray[shardIndex];
+			ShardInterval *shardInterval = list_nth(finalShardIntervalList, shardIndex);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_physical_planner.c]: Note that this is an `O(n)` operation done `n` times, making this an `O(N^2)` operation. This should just be a `foreach()` loop.

File: src/backend/distributed/utils/metadata_cache.c [hunk 17]
```diff
@@ -280,6 +302,53 @@ LookupDistTableCacheEntry(Oid relationId)
...
+			HasUniformHashDistribution(sortedShardIntervalArray,
+									   shardIntervalArrayLength);
+	}
+
+	/* check if there exists any shard intervals with no min/max values */
+	hasUninitializedShardInterval =
+		HasUninitializedShardInterval(sortedShardIntervalArray, shardIntervalArrayLength);
+
+	/* we only need hash functions for hash distributed tables */
+	if (partitionMethod == DISTRIBUTE_BY_HASH)
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: Can we collapse this and the previous DISTRIBUTE_BY_HASH block? Also, is it possibly worthwhile to move the code to a separate function?

File: src/backend/distributed/utils/metadata_cache.c [hunk 18]
```diff
@@ -298,14 +367,152 @@ LookupDistTableCacheEntry(Oid relationId)
...
 
 
 /*
+ * ShardIntervalCompareFunction returns the appropriate compare function for the
+ * partition column type. In case of hash-partitioning, it always returns the compare
+ * function for integers. Callers of this function has to ensure that shardIntervalArray
+ * has at least one element.
+ */
+static FmgrInfo *
+ShardIntervalCompareFunction(ShardInterval **shardIntervalArray, char partitionMethod)
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: I think it'd look very mildly better if these were in shardinterval_utils.c/h. But really only midlly.

**Comment by anarazel** [comment]
Conditional on that `O(n^2)` issue being fixed, you have a :shipit: . You can decide on your own whether you find the other things worth addressing or not.

File: src/backend/distributed/utils/metadata_cache.c (hunk 18 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: My idea here was to set this function as `static` and allow accesses to `shardIntervalCompareFunction` via the cache. So, I feel like keeping it as it is should be fine.

File: src/backend/distributed/utils/metadata_cache.c (hunk 17 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: Yeah, that makes sense. I collapsed both `DISTRIBUTE_BY_HASH ` into a single if.

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 16 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: Oh, good catch. Fixing it.

**Comment by onderkalaci** [comment]
Before merging in, I want to share a performance test results with the final code as a reference. The below results shows the characteristics of this change: "Shard count doesn't have a **_*very considerable* _**effect on the insert performance anymore."

Master Instance: c4.8xlarge
Table Schema: 10 integers, no indxes
Worker Instance (2): c4.xlarge + sync_commit = on
Replication factor: 1
Concurrent Insert Threads: 256
Each Test: 100 seconds

shard count | Inserts per second
------------- | ---------------
16 | 50555
64| 47726
256 | 47207
512| 48422
1024 | 49253

**Comment by anarazel** [comment]
Hi,

Nice results!

> Worker Instance (2): c4.xlarge + sync_commit = on

I suspect in this benchmark you're hitting disk bottlenecks, due to
synchronous commit = on. If the bottleneck is elsewhere, other
scalability issues don't have a large effect.

> shard count | Inserts per second
> ------------- | ---------------
> 16 | 50555
> 64| 47726
> 256 | 47207
> 512| 48422
> 1024 | 49253

Given some cucstomer questions around rebalancing I think it might be
worthwhile to repeat the test for 4096 and 16384 shards.
